### ISSUE TITLE
How to remove a channel in a keras model

### ISSUE THREAD

**Issue by dslate** [rstudio/keras#564]
I'm using keras 2.1.6 with tensorflow 1.8.0 via R 3.5.1 running under Ubuntu Linux 16.04.4.  Suppose that inside a model I have a layer that produces a 2d image tensor with shape 128,128,3.  Is there a layer that would take that as input and just remove one of the channels, producing an output with shape 128,128,2?  I presume I could do that with a layer_lambda, but I need to save and reload my models,
and as far as I can tell, keras doesn't know how to serialize a layer_lambda.

Thanks.

**Comment by skeydan** [comment]
As for `layer_lambda`, you could still use `save_model_weights_hdf5`, then recreate the model in code and  use `load_model_weights_hdf5` to restore the weights.

But perhaps if you want an output of shape (height, width, 2) you could just append a `layer_conv_2d` with `filters = 2` at the end?

**Comment by dslate** [comment]
Thanks skeydan for your response.

The problem with saving and restoring only the weights and then recreating the model in code  is that I save several models, reload them later in another run, use them all to predict for test samples, then average (ensemble) the predictions.  At that point the models are just "black boxes".  They all accept the same input and produce the same shape outputs, but their inner structures are unknown.

I did think of using a 1x1 layer_conv2d with a reduced number of filters, but it would still contain trainable weights that would average pixel values across channels, which I don't want.  I would have to set the layer weights to ignore the one or more channels I want to drop, then freeze them.  This all sounds very hacky.  Also, the layer I want is not the final output layer of the model, but occurs in the interior of the model on one of its branches.

I'm not yet very experienced using keras for image processing (in this case segmentation), so perhaps I'm just confused, and what I'm looking for is not something I should ever need. Is that the case, or is the absence of a layer that drops channels, which seems like a rather simple operation, a minor oversight by the keras developers in what seems to be a very capable and versatile neural net software package?

**Comment by dslate** [comment]
By the way, it seems to me that the ability to save and reload models containing layer_lambda would be quite useful.  Is the serialization problem really intractable, or are the keras developers working on a way to convert lambda function code into a form that can be serialized (or, in Python terminology, "pickled")?

**Comment by skeydan** [comment]
Hi,

1. if you're doing segmentation, have you seen we have a U-Net example (https://github.com/rstudio/keras/blob/master/vignettes/examples/unet.R)? Also, are you sure you need 2 filters in the output?
(As opposed to a sigmoid differentiating between on/off)?

2. Are you sure you need to keep those exact 3 channels (that is, 2 of them)? Normally things have been mixed around already quite a bit (assuming you've already done a few convolutions before). I'd just try the 1x1 conv with 2 filters. You could experiment with ```pointwise_regularizer``` or ```pointwise_constraint``` in a separable convolution though, to diminish the effect of channel mixing. See ```?layer_separable_conv_2d```.

3. I don't see why you couldn't regenerate the model programatically - even different models - at a later time?

4. For saving, if you wanted to use eager execution you could use the TensorFlow way of saving weights ("checkpointing"): see https://tensorflow.rstudio.com/keras/articles/eager_guide.html#saving-and-restoring-model-weights

**Comment by skeydan** [comment]
Thanks for the explanations.

> I still think, however, that a layer that discards specified channels could be considered a fundamental operation that perhaps ought to be implemented in keras. I wonder if I am the only user that has ever desired this facility, or whether there might be others who would find it useful.

You might want to ask this on Python Keras github :-)
